File: src/core/engine/inputs.ts
/**
 * Inputs, resolved once at load, `compiled-program.md` 2.6 and 5.1.
 *
 * Input resolution and the substitution of every `{ "input": ... }` reference
 * happen **once, before step 1 of bar 0**, so the declarations in `outputs` and
 * anything built from them exist before the first bar runs. Step 5 writes
 * already resolved values into slots and resolves nothing.
 *
 * **A value that fails validation is a refusal, not a fallback.** The host's
 * value is used when it passes and the declared default when the host supplies
 * none, and a supplied value that fails is OS6019 and the program does not run.
 * Falling back to the default would be a settings dialog that silently ignores
 * what a user typed, which is worse than one that says the value is out of
 * range.
 *
 * **An input value is never absent.** `input()` cannot declare `none` as a
 * default, because the default's type is what fixes the input's type.
 */
import type { Diagnostic } from '../diagnostics/index.js';
import { isBarField } from './bars.js';
import { NO_POSITION, failure } from './errors.js';
import type { CompiledInput, CompiledProgram, Constant, Field } from './types.js';
import type { Value } from './values/index.js';

/** The eight built-in series a `"source"` input may select, 2.6. */
const SOURCES: readonly string[] = [
  'open',
  'high',
  'low',
  'close',
  'hl2',
  'hlc3',
  'ohlc4',
  'volume',
];

export interface ResolvedInput {
  readonly key: string;
  readonly slot: number;
  /** The bar field a `"source"` input selected, which step 5 reads per bar. */
  readonly field: string | undefined;
  /** The effective value, for every other kind. */
  readonly value: Value;
}

export type InputResult =
  | { readonly ok: true; readonly inputs: readonly ResolvedInput[] }
  | { readonly ok: false; readonly diagnostic: Diagnostic };

/** A host's own conversion for a `"time"` input, which needs the chart's zone. */
export type TimeResolver = (text: string) => number | null;

/** The value a constant pool entry denotes. */
export function constantValue(entry: Constant): Value {
  switch (entry[0]) {
    case 'z':
      return null;
    case 'b':
    case 'n':
    case 's':
      return entry[1];
    default:
      return { tag: 'color', r: entry[1][0], g: entry[1][1], b: entry[1][2], a: entry[1][3] };
  }
}

export function resolveInputs(
  program: CompiledProgram,
  settings: Readonly<Record<string, unknown>>,
  resolveTime: TimeResolver,
): InputResult {
  const inputs: ResolvedInput[] = [];
  for (const declared of program.inputs) {
    const resolved = resolveOne(declared, settings, resolveTime);
    if ('diagnostic' in resolved) return { ok: false, diagnostic: resolved.diagnostic };
    inputs.push(resolved.input);
  }
  return { ok: true, inputs };
}

function refuse(key: string, value: unknown, validation: string): {
  readonly diagnostic: Diagnostic;
} {
  return {
    diagnostic: failure('OS6019', NO_POSITION, { value: describe(value), key, validation }),
  };
}

function resolveOne(
  declared: CompiledInput,
  settings: Readonly<Record<string, unknown>>,
  resolveTime: TimeResolver,
): { readonly input: ResolvedInput } | { readonly diagnostic: Diagnostic } {
  const supplied = Object.prototype.hasOwnProperty.call(settings, declared.key)
    ? settings[declared.key]
    : undefined;
  const fallback = constantValue(declared.default);

  if (declared.kind === 'source') {
    const field = supplied === undefined ? fallback : supplied;
    if (typeof field !== 'string' || !SOURCES.includes(field)) {
      return refuse(declared.key, field, `a source names one of ${SOURCES.join(', ')}`);
    }
    if (!isBarField(field)) {
      return refuse(declared.key, field, 'this engine has no bar field of that name');
    }
    return { input: { key: declared.key, slot: declared.slot, field, value: null } };
  }

  if (declared.kind === 'time') {
    const held = supplied === undefined ? fallback : supplied;
    const time = typeof held === 'number' ? held : typeof held === 'string' ? resolveTime(held) : null;
    if (time === null) {
      return refuse(declared.key, held, 'a time is a timestamp or a date and time the host can read');
    }
    return { input: { key: declared.key, slot: declared.slot, field: undefined, value: time } };
  }

  if (supplied === undefined) {
    return { input: { key: declared.key, slot: declared.slot, field: undefined, value: fallback } };
  }

  const wrong = validate(declared, supplied);
  if (wrong !== undefined) return refuse(declared.key, supplied, wrong);
  return {
    input: { key: declared.key, slot: declared.slot, field: undefined, value: supplied as Value },
  };
}

/** What rule the host's value broke, or nothing when it passes. */
function validate(declared: CompiledInput, supplied: unknown): string | undefined {
  switch (declared.kind) {
    case 'number': {
      if (typeof supplied !== 'number' || !Number.isFinite(supplied)) {
        return 'this input takes a number';
      }
      if (declared.min !== null && supplied < declared.min) {
        return `the minimum is ${declared.min}`;
      }
      if (declared.max !== null && supplied > declared.max) {
        return `the maximum is ${declared.max}`;
      }
      return undefined;
    }
    case 'bool':
      return typeof supplied === 'boolean' ? undefined : 'this input takes true or false';
    case 'color':
      return isColourValue(supplied) ? undefined : 'this input takes a colour';
    case 'select': {
      if (typeof supplied !== 'string') return 'this input takes one of its listed values';
      const allowed = (declared.options ?? []).map((one) => constantValue(one));
      if (!allowed.includes(supplied)) {
        return `the choices are ${allowed.map((one) => String(one)).join(', ')}`;
      }
      return undefined;
    }
    default:
      return typeof supplied === 'string' ? undefined : 'this input takes a string';
  }
}

function isColourValue(value: unknown): boolean {
  if (typeof value !== 'object' || value === null) return false;
  const held = value as Record<string, unknown>;
  return (
    held['tag'] === 'color' &&
    typeof held['r'] === 'number' &&
    typeof held['g'] === 'number' &&
    typeof held['b'] === 'number' &&
    typeof held['a'] === 'number'
  );
}

/**
 * A declaration field with its `{ "input": key }` reference substituted.
 *
 * Every field of `meta`, of `meta.strategy` and of every declaration in
 * `outputs` may hold the reference, and it is gone before bar 0.
 */
export function fieldValue(field: Field, inputs: readonly ResolvedInput[]): Value {
  if (field === null) return null;
  if (typeof field === 'object' && !Array.isArray(field)) {
    const key = (field as { readonly input: string }).input;
    const found = inputs.find((one) => one.key === key);
    return found === undefined ? null : found.value;
  }
  if (Array.isArray(field)) {
    const [r, g, b, a] = field as readonly number[];
    if (r === undefined || g === undefined || b === undefined || a === undefined) return null;
    return { tag: 'color', r, g, b, a };
  }
  return field as Value;
}

/**
 * A date and time read as if it were UTC.
 *
 * **This is a stand-in.** `stdlib.md` 12.1 reads every calendar field in the
 * chart's timezone, which is an IANA zone name the host states, and an engine
 * that has not been given one cannot apply it. A host with a zone supplies its
 * own resolver; this one is what a test and a host with no zone get, and it is
 * deterministic, which is the property that matters most here.
 */
const STAMP = /^(\d{4})-(\d{2})-(\d{2})(?:[ T](\d{2}):(\d{2})(?::(\d{2}))?)?$/;

export function utcTime(text: string): number | null {
  const parsed = STAMP.exec(text.trim());
  if (parsed === null) return null;
  const at = (index: number): number => Number(parsed[index] ?? '0');
  const value = Date.UTC(at(1), at(2) - 1, at(3), at(4), at(5), at(6));
  return Number.isFinite(value) ? value : null;
}

function describe(value: unknown): string {
  if (value === undefined || value === null) return 'none';
  if (typeof value === 'string') return JSON.stringify(value);
  if (typeof value === 'number' || typeof value === 'boolean') return String(value);
  return 'a value of another type';
}

File: engine/openscript/inputs.py
"""Inputs, resolved once at load, sections 2.6 and 5.1.

Input resolution and the substitution of every ``{ "input": ... }`` reference
happen **once, before step 1 of bar 0**, so the declarations in ``outputs`` and
anything built from them exist before the first bar runs. Step 5 writes already
resolved values into slots and resolves nothing.

**A value that fails validation is a refusal, not a fallback.** The host's value
is used when it passes and the declared default when the host supplies none, and
a supplied value that fails is OS6019 and the program does not run. Falling back
to the default would be a settings dialog that silently ignores what a user
typed, which is worse than one that says the value is out of range.

**An input value is never absent.** ``input()`` cannot declare ``none`` as a
default, because the default's type is what fixes the input's type.
"""

from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any, Callable, Dict, List, Mapping, Optional, Sequence, Tuple

from .bars import is_bar_field
from .diagnostics import Diagnostic, failure
from .program import constant_value
from .values import ABSENT, Colour, is_number

#: The eight built-in series a ``"source"`` input may select, section 2.6.
SOURCES: Tuple[str, ...] = (
    "open",
    "high",
    "low",
    "close",
    "hl2",
    "hlc3",
    "ohlc4",
    "volume",
)


@dataclass(frozen=True)
class ResolvedInput:
    key: str
    slot: int
    #: The bar field a ``"source"`` input selected, which step 5 reads per bar.
    field: Optional[str]
    #: The effective value, for every other kind.
    value: Any


@dataclass(frozen=True)
class InputResult:
    inputs: Sequence[ResolvedInput]
    diagnostic: Optional[Diagnostic] = None

    @property
    def ok(self) -> bool:
        return self.diagnostic is None


TimeReader = Callable[[str], Optional[float]]


def resolve_inputs(
    raw: Mapping[str, Any],
    settings: Mapping[str, Any],
    read_time: Optional[TimeReader] = None,
) -> InputResult:
    found: List[ResolvedInput] = []
    for declared in raw["inputs"]:
        one, refusal = _resolve_one(declared, settings, read_time)
        if refusal is not None:
            return InputResult((), refusal)
        found.append(one)
    return InputResult(tuple(found))


def _resolve_one(
    declared: Mapping[str, Any],
    settings: Mapping[str, Any],
    read_time: Optional[TimeReader],
) -> Tuple[Optional[ResolvedInput], Optional[Diagnostic]]:
    supplied = settings[declared["key"]] if declared["key"] in settings else None
    fallback = constant_value(declared["default"])

    # A source and a time are checked even where the host stored nothing,
    # because each declares its value as text this engine still has to read.
    # Every other kind takes its declared default unexamined: a default is
    # written in the source, and validation here is about a value that arrived
    # from outside it.
    ask_anyway = declared["kind"] in ("source", "time")
    if declared["key"] not in settings and not ask_anyway:
        return ResolvedInput(declared["key"], declared["slot"], None, fallback), None

    held = fallback if declared["key"] not in settings else supplied
    value, field, refusal = check_setting(declared, held, read_time)
    if refusal is not None:
        return None, failure(
            "OS6019", value=_describe(held), key=declared["key"], validation=refusal
        )
    return ResolvedInput(declared["key"], declared["slot"], field, value), None


def check_setting(
    declared: Mapping[str, Any],
    supplied: Any,
    read_time: Optional[TimeReader] = None,
) -> Tuple[Any, Optional[str], Optional[str]]:
    """What the engine will run with for one stored value, or the rule it breaks.

    Answered outside the engine as well: a chart builds a settings dialog before
    anything is loaded, so it has to know what a stored value is worth before the
    run that would refuse it exists. A second copy of these rules in an adapter
    would be the same fact in two files, answering differently the first time
    either was edited.
    """
    kind = declared["kind"]
    if kind == "source":
        if not isinstance(supplied, str) or supplied not in SOURCES:
            return None, None, f"a source names one of {', '.join(SOURCES)}"
        if not is_bar_field(supplied):
            return None, None, "this engine has no bar field of that name"
        return ABSENT, supplied, None

    if kind == "time":
        unreadable = "a time is a timestamp or a date and time the host can read"
        if is_number(supplied):
            return float(supplied), None, None
        if not isinstance(supplied, str):
            return None, None, unreadable
        if read_time is None:
            return supplied, None, None
        when = read_time(supplied)
        if when is None:
            return None, None, unreadable
        return float(when), None, None

    broken = _validate(declared, supplied)
    if broken is not None:
        return None, None, broken
    return _as_value(supplied), None, None


def _as_value(supplied: Any) -> Any:
    """A stored setting as a machine value: a number is a float, a colour a colour."""
    if isinstance(supplied, bool) or isinstance(supplied, str) or supplied is None:
        return supplied if supplied is not None else ABSENT
    if is_number(supplied):
        return float(supplied)
    if isinstance(supplied, Colour):
        return supplied
    if isinstance(supplied, (list, tuple)) and len(supplied) == 4:
        red, green, blue, alpha = supplied
        return Colour(float(red), float(green), float(blue), float(alpha))
    return supplied


def _validate(declared: Mapping[str, Any], supplied: Any) -> Optional[str]:
    """What rule the host's value broke, or nothing when it passes."""
    kind = declared["kind"]
    if kind == "number":
        if not is_number(supplied):
            return "this input takes a number"
        if declared.get("min") is not None and supplied < declared["min"]:
            return f"the minimum is {declared['min']}"
        if declared.get("max") is not None and supplied > declared["max"]:
            return f"the maximum is {declared['max']}"
        return None
    if kind == "bool":
        return None if isinstance(supplied, bool) else "this input takes true or false"
    if kind == "color":
        return None if _is_colour(supplied) else "this input takes a colour"
    if kind == "select":
        if not isinstance(supplied, str):
            return "this input takes one of its listed values"
        allowed = [constant_value(one) for one in (declared.get("options") or [])]
        if supplied not in allowed:
            return f"the choices are {', '.join(str(one) for one in allowed)}"
        return None
    return None if isinstance(supplied, str) else "this input takes a string"


def _is_colour(value: Any) -> bool:
    if isinstance(value, Colour):
        return True
    return isinstance(value, (list, tuple)) and len(value) == 4 and all(
        is_number(one) for one in value
    )


def field_value(field: Any, inputs: Sequence[ResolvedInput]) -> Any:
    """A declaration field with its ``{ "input": key }`` reference substituted.

    Every field of ``meta``, of ``meta.strategy`` and of every declaration in
    ``outputs`` may hold the reference, and it is gone before bar 0.
    """
    if field is None:
        return ABSENT
    if isinstance(field, dict):
        key = field.get("input")
        for one in inputs:
            if one.key == key:
                return one.value
        return ABSENT
    if isinstance(field, (list, tuple)) and len(field) == 4 and all(is_number(one) for one in field):
        red, green, blue, alpha = field
        return Colour(float(red), float(green), float(blue), float(alpha))
    return field


#: A date and time read as if it were UTC.
#:
#: **A stand-in.** ``stdlib.md`` section 12.1 reads every calendar field in the
#: chart's timezone, which is an IANA name the host states, and an engine that
#: has not been given one cannot apply it. A host with a zone supplies its own
#: reader; this is what a test and a host with no zone get, and it is
#: deterministic, which is the property that matters most here.
def utc_time(text: str) -> Optional[float]:
    for pattern in ("%Y-%m-%d %H:%M:%S", "%Y-%m-%dT%H:%M:%S", "%Y-%m-%d %H:%M",
                    "%Y-%m-%dT%H:%M", "%Y-%m-%d"):
        try:
            when = datetime.strptime(text.strip(), pattern)
        except ValueError:
            continue
        return when.replace(tzinfo=timezone.utc).timestamp() * 1000
    return None


def _describe(value: Any) -> str:
    if value is None:
        return "none"
    if isinstance(value, str):
        return f'"{value}"'
    if isinstance(value, bool) or is_number(value):
        return str(value)
    return "a value of another type"


def settings_slots(inputs: Sequence[ResolvedInput]) -> Dict[int, ResolvedInput]:
    return {one.slot: one for one in inputs}
